compiler/env/CompileTimeProfiler.hpp

```diff
@@ -0,0 +1,105 @@
+/*******************************************************************************
+ *
+ * (c) Copyright IBM Corp. 2017, 2017
+ *
+ *  This program and the accompanying materials are made available
+ *  under the terms of the Eclipse Public License v1.0 and
+ *  Apache License v2.0 which accompanies this distribution.
+ *
+ *      The Eclipse Public License is available at
+ *      http://www.eclipse.org/legal/epl-v10.html
+ *
+ *      The Apache License v2.0 is available at
+ *      http://www.opensource.org/licenses/apache2.0.php
+ *
+ * Contributors:
+ *    Multiple authors (IBM Corp.) - initial implementation and documentation
+ ******************************************************************************/
+
+#ifndef OMR_COMPILE_TIME_PROFILER_HPP
+#define OMR_COMPILE_TIME_PROFILER_HPP
+
+#include "compile/Compilation.hpp"
+#include <fcntl.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <sys/types.h>
+
+#include <sys/syscall.h>
+#include <stdio.h>
+
+namespace TR {
+
+/**
+ * Utility to generate a perf profile for the current compilation,
+ * to assist in determining the source of compile time overheah. This
+ * is useful in situations where there is a particularly long running
+ * compile, such as a profiled very-hot or scorching compile, but profiling
+ * all compilations results in too much noise.
+ *
+ * This will fork and execute perf, configured such that it will
+ * profile the current compilation thread until a time limit
+ * is met or the current compilation ends.
+ *
+ * Perf is configured to collect call graph information using frame
+ * pointers at a high frequency, using the options:
+ *   --call-graph fp -m 4096 -F9999 -r 1
```

Comment:
Perhaps not a great idea to hard code some of these options, it will force people to edit the code and rebuild if they ever try this in an environment where the defaults don't work.

Regarding `--call-graph fp`, did you test this on something other than AMD64 or did you build with frame pointers enabled? This method of call stack sampling basically doesn't work on AMD64 unless you build your code with FP support, you'll just get bogus results; `--call-graph dwarf` is the more reliable option, but may not work on anything other than AMD64 unfortunately.

`-r` needs root or the user needs to have the right caps IIRC, and the specific `-m` and `-F` values you're using can fail you on some systems depending on how they're set up.

Reply:
>Perhaps not a great idea to hard code some of these options, it will force people to edit the code and rebuild if they ever try this in an environment where the defaults don't work.

You could probably accept a string from the user, similar to  how it's done with the counts string option:

https://github.com/eclipse/omr/blob/28966b753f1c58bb25d8a56a311dc1cb065138e3/compiler/control/OMROptions.cpp#L192-L197

Reply:
I know from having been working with @ncough that we did do runs where we build the JIT with the frame pointers enabled so we got accurate profiling - hence the option. I agree either an option or an env var to set the string is worth while. These defaults have been very helpful so maybe they can be the default string if none is supplied.

Reply:
The implementation has been extended to accept arguments through an environment variable, with a set of sane defaults. 

@ymanton You are right, it was necessary to build with frame pointers enabled, as the particular machines tested at the time did not support dwarf. However, as this should be the exception, the default has been changed to dwarf.